### DISCUSSION TITLE
Android-Client doesn't connect with IPV6-only-server

### DISCUSSION THREAD

**Discussion by hp4** [strongswan/strongswan#1694] [Q&A]
Hi,

does anybody know about IPV6-name-resolution-problems on Android-devices? I can't figure out, why my client doesn't connect with IPV6-only-Strongswan-server. Is it an Android-OS bug? Or a bug in the Strongswan-Android-client? Or a misconfiguration on my side? Details see below

Thanks for any help or suggestions.
BR
Hans-Peter 

I've 2 Strongswan-Servers, one for IPV4 and the second with IPV6-address only.
With the IPV4-Server everything works fine (connections from Windows/Linux/Android-clients).

The IPV6-only-Server works fine with Windows-clients!!!

But I can't get a working connection with the Android-client: 

- without enforced UDP-encapsulation the client refuses to connect right from the beginning:

May 12 15:58:08 08[CFG] loaded CA certificate .....
May 12 15:58:08 17[LIB] resolving 'xxxvpn.dynv6.net' failed: No address associated with hostname
May 12 15:58:08 08[IKE] unable to resolve xxxvpn.dynv6.net, initiate aborted
May 12 15:58:08 08[CFG] failed to initiate tunnel

When the FQDN is replaced by the corresponding IPV6-address I get the same error

May 12 15:56:56 14[CFG] loaded CA certificate ...
May 12 15:56:56 14[IKE] unable to resolve 2001:9e8:8c9f:500:ba27:ebff:fe75:37b4, initiate aborted
May 12 15:56:56 14[CFG] failed to initiate tunnel

- when UDP-encapsulation is enforced, the client is able to connect, the name resolution obviously works

May 12 15:59:07 07[CFG] loaded CA certificate ...
May 12 15:59:07 07[IKE] initiating IKE_SA android[36] to 2001:9e8:8c9f:500:ba27:ebff:fe75:37b4
May 12 15:59:07 07[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
May 12 15:59:07 07[NET] sending packet: from 2001:9e8:8c9f:501:8866:80e:1412:6ab4[57603] to 2001:9e8:8c9f:500:ba27:ebff:fe75:37b4[500] (948 bytes)
May 12 15:59:07 10[NET] received packet: from 2001:9e8:8c9f:500:ba27:ebff:fe75:37b4[500] to 2001:9e8:8c9f:501:8866:80e:1412:6ab4[57603] 

but eventually the connection gets destroyed because of "NETLINK ERROR 22", which is probably due to missing support for UDP-encapsulation in older Linux kernels.

**Comment by tobiasbrunner** [comment]
> * without enforced UDP-encapsulation

What do you mean by that? The Android app doesn't have such an option as it always forces UDP encapsulation (it doesn't work any other way as it can only send ESP packets via IKE UDP socket). And forcing UDP encapsulation on the server obviously has no effect on the client at that point.

Are you perhaps referring to the IPv6 option in the advanced profile settings? If so, you definitely have to enable that.

> but eventually the connection gets destroyed because of "NETLINK ERROR 22",

Again, what are you referring to? The Android client doesn't access Netlink. Or do you get that on the server? If so, yes, this won't work on a kernel that doesn't support UDP encapsulation for IPv6 as, again, the Android client requires that (that's why the IPv6 option is disabled in the app, by default).

**Reply by hp4** [reply]
Thanks for your comment, sorry for being late with my answer.

what does "unable to resolve xxxvpn.dynv6.net, initiate aborted" mean, when the client immediately returns? 

That's  a misleading error-message. DNS-resolution of the address is definitly possible! This is obvious, because the client continues with UDP-encapsulation enabled in the client options.

In the latter case the tunnel can not be established and the strongswan-server.log-file reports the NETLINK error. I'm not an expert and googled a lot to find the problem. And eventually I found several hints that UDP-encapsulation ist only supported by most recent Linux-kernel and strongswan versions. So I disabled UDP-encapsulation.

I've tried to read all your relevant IPV6-documentation but maybe I missed something?

Does your answer mean: Android clients won't work with older Linux-kernel/strongswan-versions in IPV6? System-requirement?

Where can I find your ".... you definitely have to enable that." in your online-docs? And why do you provide this option at all, if "enabled" is the only allowed value?

Maybe I misinterpreted your docs in this way: you can use servers with older Linux-kernels as long as you don't use UDP-encapsulation? 

Thanks for your help

**Reply by tobiasbrunner** [reply]
> That's a misleading error-message. DNS-resolution of the address is definitly possible! This is obvious, because the client continues with UDP-encapsulation enabled in the client options.

Again, there is no such option. However, there is an option to enable IPv6 support ("Use IPv6 transport addresses"). And only if that's enabled can IPv6 addresses be resolved. Otherwise, DNS resolution is strictly limited to IPv4 and if your domain has no A records that will fail with the error message you saw.

> And eventually I found several hints that UDP-encapsulation ist only supported by most recent Linux-kernel and strongswan versions.

The strongSwan version doesn't matter. But that there could be issues with the server's kernel version is also clearly stated in the description of the IPv6 option in the app itself. It says: "Can only be enabled if UDP encapsulation for IPv6 is supported by the server. Note that the Linux kernel only supports this since version 5.8, so many servers will not support it yet."

> I've tried to read all your relevant IPV6-documentation but maybe I missed something?

Seems that way, because I don't see what could be unclear.

> Does your answer mean: Android clients won't work with older Linux-kernel/strongswan-versions in IPV6?

Yes, exactly.

> Where can I find your ".... you definitely have to enable that." in your online-docs? And why do you provide this option at all, if "enabled" is the only allowed value?

What do you mean? If you want to use IPv6, you have to enable the option for it in the app. It's disabled exactly because of the issue you ran into with older Linux kernels that don't support UDP-encapsulation for IPv6.

> Maybe I misinterpreted your docs in this way: you can use servers with older Linux-kernels as long as you don't use UDP-encapsulation?

With strongSwan in general you can do that, but not with the Android app because it **requires** UDP-encapsulation (it doesn't have permission to send raw ESP packets).

It's really quite simple: The app only has permission to send UDP packets, so it always enforces UDP-encapsulation for ESP. This means that UDP-encapsulation for IPv6 is required to use that address family. But because some servers might not (e.g. older Linux kernels) and the app has no way to determine that, the option to use IPv6 is disabled by default and should only be enabled by the user when the server is verified to be compatible.

**Reply by hp4** [reply]
Ok, thanks a lot for your detailed answer.
Would it be possible to state these restrictions in the "Client configuration"-section of 
https://docs.strongswan.org/docs/5.9/os/androidVpnClient.html? For an unexperienced user like me that would be the expected location.
 
In the meantime I've found this information in the Changlog statements 1.3.0 and 2.3.1 at the end of that page, but it takes much more time to study all the changelogs.

**Reply by gr4yg00s3** [reply]
>  `May 12 15:58:08 17[LIB] resolving 'xxxvpn.dynv6.net' failed: No address associated with hostname`

I see this while trying to setup a v6 only net2net tunnel, is there something I'm missing or is this expected?
